ql/src/java/org/apache/hadoop/hive/ql/optimizer/GenMapRedUtils.java

```diff
@@ -1309,12 +1312,29 @@ public static void createMRWorkForMergingFiles(FileSinkOperator fsInput,
     // 2. Constructing a conditional task consisting of a move task and a map reduce task
     //
     Path inputDirName = fsInputDesc.getMergeInputDirName();
+
+    // For external CTAS queries, there is an optimisation which avoids rename/copy of
+    // files from -tmp to -ext. In such cases, the input dir must be -tmp.
+    boolean changeToTmpPath = false;
+    FileSystem fs;
+    try {
+      fs = inputDirName.getFileSystem(conf);
+      if (fsInputDesc.isCTASorCM() && fsInputDesc.getTable().getTableType().equals(TableType.EXTERNAL_TABLE)
+              && BlobStorageUtils.isBlobStorageFileSystem(conf, fs)
+              && !Utilities.shouldAvoidRename(fsInputDesc, conf)) {
+        changeToTmpPath = true;
+      }
+    } catch (IOException e) {
+      throw new RuntimeException("Unable to get filesystem from path due to: ", e);
```

Comment:
Is this a common practice to pack the IOException to RuntimeException?

Reply:
Nice catch, updated to `SemanticException`. I think the idea is to convey compile time exceptions in SemanticException.